demo/parser_demo.cpp

```diff
@@ -129,14 +129,61 @@ class PrintVisitor : public ast::Visitor {
             depth--;
         }
     };
-    void visit(ast::Expression* node) override{
-
+    void visit(ast::Expression* node) override {
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [Expression]> null\n", "", depth);
+        } else {
+            // if binary
+            if (node->operand2) {
+
+                fmt::print("{:|>{}}- [BinaryOperator]> {}\n", "", depth,
+                           node->operation.lit);
+                depth++;
+                node->operand1->accept(*this);
+                node->operand2->accept(*this);
+                depth--;
+            } else {
+
+                fmt::print("{:|>{}}- [UnaryOperator]> {}\n", "", depth,
+                           node->operation.lit);
+                depth++;
+                node->operand1->accept(*this);
+                depth--;
+            }
+        }
     };
-    void visit(ast::UnaryExpression* node) override{
-
+    void visit(ast::Primitive* node) override {
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [Primitive]> null\n", "", depth);
+        } else {
+            fmt::print("{:|>{}}- [Primitive]> {}\n", "", depth,
+                       node->value.lit);
+        }
     };
-    void visit(ast::BinaryExpression* node) override{
-
+    void visit(ast::ModifiablePrimary* node) override {
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [ModifiablePrimary]> null\n", "", depth);
+        } else {
+            fmt::print("{:|>{}}- [ModifiablePrimary]> size {} \n", "", depth,
+                       node->args.size());
+            depth++;
+            for (int i = 0; i < node->args.size(); i++) {
+                node->args[i]->accept(*this);
+            }
+        }
+        depth--;
+    };
+    void visit(ast::RoutineCall* node) override {
+        if (node == nullptr) {
+            fmt::print("{:|>{}}- [RoutineCall]> null\n", "", depth);
+        } else {
+            fmt::print("{:|>{}}- [RoutineCall]> {}\n", "", depth,
+                       node->routine.lit);
+            depth++;
+            for (int i = 0; i < node->args.size(); i++) {
+                node->args[i]->accept(*this);
+            }
+        }
```

Comment:
I thought we fixed those strange `depth++` that don't have a matching `depth--` in the same level